Mathlib/NumberTheory/LSeries/Nonvanishing.lean

```diff
@@ -0,0 +1,241 @@
+private lemma re_log_comb_nonneg {n : ℕ} (hn : 2 ≤ n) {x : ℝ} (hx : 1 < x) (y : ℝ) :
+    0 ≤ 3 * (-log (1 - (1 : DirichletCharacter ℂ N) n * n ^ (-x : ℂ))).re +
+          4 * (-log (1 - χ n * n ^ (-(x + I * y)))).re +
+          (-log (1 - (χ n ^ 2) * n ^ (-(x + 2 * I * y)))).re := by
+  by_cases hn' : IsUnit (n : ZMod N)
+  · have ha₀ : 0 ≤ (n : ℝ) ^ (-x) := Real.rpow_nonneg n.cast_nonneg _
+    have ha₁ : (n : ℝ) ^ (-x) < 1 := by
+      rw [Real.rpow_neg (Nat.cast_nonneg n), inv_lt_one_iff₀]
+      exact .inr <| Real.one_lt_rpow (mod_cast one_lt_two.trans_le hn) <| zero_lt_one.trans hx
+    have hz : ‖χ n * (n : ℂ) ^ (-(I * y))‖ = 1 := by
+      rw [norm_mul, ← hn'.unit_spec, DirichletCharacter.unit_norm_eq_one χ hn'.unit,
+        norm_eq_abs, ← ofReal_natCast, abs_cpow_eq_rpow_re_of_pos (mod_cast by omega)]
+      simp only [neg_re, mul_re, I_re, ofReal_re, zero_mul, I_im, ofReal_im, mul_zero, sub_self,
+        neg_zero, Real.rpow_zero, one_mul]
+    rw [MulChar.one_apply hn', one_mul]
+    convert _root_.re_log_comb_nonneg ha₀ ha₁ hz using 6
+    · congr 2
+      exact_mod_cast (ofReal_cpow n.cast_nonneg (-x)).symm
+    · congr 2
+      rw [neg_add, cpow_add _ _ <| mod_cast by omega, ← ofReal_neg, ofReal_cpow n.cast_nonneg (-x),
+        ofReal_natCast, mul_left_comm]
+    · rw [neg_add, cpow_add _ _ <| mod_cast by omega, ← ofReal_neg, ofReal_cpow n.cast_nonneg (-x),
+        ofReal_natCast, show -(2 * I * y) = (2 : ℕ) * -(I * y) by ring, cpow_nat_mul, mul_pow,
+        mul_left_comm]
+  · simp only [MulChar.map_nonunit _ hn', zero_mul, sub_zero, log_one, neg_zero, zero_re, mul_zero,
+      neg_add_rev, add_zero, ne_eq, OfNat.ofNat_ne_zero, not_false_eq_true, zero_pow, le_refl]
+
+/-- The logarithms of the Euler factors of a Dirichlet L-series form a summable sequence. -/
+lemma summable_neg_log_one_sub_mul_prime_cpow {s : ℂ} (hs : 1 < s.re) :
+    Summable fun p : Nat.Primes ↦ -log (1 - χ p * (p : ℂ) ^ (-s)) := by
+  have (p : Nat.Primes) : ‖χ p * (p : ℂ) ^ (-s)‖ ≤ (p : ℝ) ^ (-s).re := by
+    rw [norm_mul, norm_natCast_cpow_of_re_ne_zero _ <| re_neg_ne_zero_of_one_lt_re hs]
+    conv_rhs => rw [← one_mul (_ ^ _)]
+    gcongr
+    exact DirichletCharacter.norm_le_one χ _
+  refine (Nat.Primes.summable_rpow.mpr ?_).of_nonneg_of_le (fun _ ↦ norm_nonneg _) this
+    |>.of_norm.clog_one_sub.neg
+  simp only [neg_re, neg_lt_neg_iff, hs]
+
+private lemma one_lt_re_one_add {x : ℝ} (hx : 0 < x) (y : ℝ) :
+    1 < (1 + x : ℂ).re ∧ 1 < (1 + x + I * y).re ∧ 1 < (1 + x + 2 * I * y).re := by
+  simp only [add_re, one_re, ofReal_re, lt_add_iff_pos_right, hx, mul_re, I_re, zero_mul, I_im,
+    ofReal_im, mul_zero, sub_self, add_zero, re_ofNat, im_ofNat, mul_one, mul_im, and_self]
+
+open scoped LSeries.notation in
+/-- For positive `x` and nonzero `y` and a Dirichlet character `χ` we have that
+$|L(\chi^0, x)^3 \cdot L(\chi, x+iy)^4 \cdot L(\chi^2, x+2iy)| \ge 1$. -/
```

Comment:
If you hover over the lemma name, the docstring is displayed with the latex un-processed, so I think using unicode text would be better.

Reply:
Will do.